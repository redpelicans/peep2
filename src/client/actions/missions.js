import { isEmpty } from 'ramda';

export const LOAD_MISSIONS = 'EvtX:Server:missions:load';
export const MISSIONS_LOADED = 'missions:loaded';
export const MISSION_ADDED = 'mission:added';
export const ADD_MISSION = 'EvtX:Server:missions:add';

export const loadMissions = () => (dispatch, getState) => {
  const { people } = getState();
  if (isEmpty(people.data)) {
    dispatch({
      type: LOAD_MISSIONS,
      replyTo: MISSIONS_LOADED,
    });
  }
};

export const addMission = mission => dispatch => {
  dispatch({
    type: ADD_MISSION,
    payload: mission,
    replyTo: MISSION_ADDED,
  });
};
