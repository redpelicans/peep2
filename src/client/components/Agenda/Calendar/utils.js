import { Colors } from '@blueprintjs/core';
import { isValidated } from '../../../utils/events';

export const vacationDayBackground = event =>
  event && !isValidated(event)
    ? `repeating-linear-gradient(45deg, ${Colors.GREEN1}, ${Colors.GREEN1} 7%, ${Colors.GREEN3} 7%, ${Colors.GREEN3} 14%)`
    : Colors.GREEN3;
export const sickLeaveDayBackground = event =>
  event && !isValidated(event)
    ? `repeating-linear-gradient(45deg, ${Colors.RED1}, ${Colors.RED1} 7%, ${Colors.RED3} 7%, ${Colors.RED3} 14%)`
    : Colors.RED3;
export const spareDayBackground = Colors.GRAY1;
export const workingDayBackground = Colors.DARK_GRAY4;
export const selectedBackground = Colors.GREEN5;
