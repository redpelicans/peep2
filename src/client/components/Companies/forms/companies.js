const fields = {
  types: {
    key: 'Types',
    label: 'Types',
    type: 'select',
    domainValues: ['Client', 'Partner', 'Tenant'],
    row: 1,
    column: 1,
    required: true,
  },
  name: {
    key: 'Name',
    label: 'Name',
    type: 'input',
    row: 1,
    column: 2,
    required: true,
  },
  website: {
    key: 'Website',
    label: 'Website',
    type: 'input',
    row: 1,
    column: 3,
    required: true,
  },
  street: {
    key: 'Street',
    label: 'Street',
    type: 'input',
    row: 2,
    column: 1,
    required: true,
  },
  zipcode: {
    key: 'Zipcode',
    label: 'Zip Code',
    type: 'input',
    row: 2,
    column: 2,
    required: true,
  },
  city: {
    key: 'City',
    label: 'City',
    type: 'select',
    row: 3,
    column: 1,
    required: true,
  },
  country: {
    key: 'Country',
    label: 'Country',
    type: 'select',
    row: 3,
    column: 2,
    required: true,
  },
  tags: {
    key: 'Tags',
    label: 'Tags',
    type: 'select',
    row: 4,
    column: 1,
    required: false,
  },
  notes: {
    key: 'Notes',
    label: 'Notes',
    type: 'textarea',
    row: 5,
    column: 1,
    required: false,
  },
};

export default fields;
