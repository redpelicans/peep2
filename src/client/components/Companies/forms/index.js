import { compose, reduce, toPairs } from 'ramda';
import fields from './companies';

// const sortFieldsByRows = sfields => groupBy(prop('row'), sfields);
// const sortFieldsByColumns = sfields => map(row => , sfields);

const exportedFields = compose(reduce((acc, [name, r]) => [...acc, { ...r, name }], []), toPairs)(fields);
export default exportedFields;
