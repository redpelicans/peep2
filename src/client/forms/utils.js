import { filter, compose, reduce, toPairs, map } from 'ramda';

const getDomainIds = domainValues => map(x => x.id, domainValues || []);
const extendWithRequired = ({ label, required }) => rule =>
  required ? rule.required(`'${label}' is required!`) : rule;
const extendWithDomainValues = ({ domainValues }) => rule =>
  domainValues ? rule.oneOf(getDomainIds(domainValues)) : rule;
const makeValidationRule = field => {
  const { validate } = field;
  if (!validate) return;
  return reduce((acc, fn) => fn(acc), validate, [
    extendWithRequired(field),
    extendWithDomainValues(field),
  ]);
};

export const getOneValidationSchema = compose(
  reduce((acc, [name, field]) => ({ ...acc, [name]: field.validate }), {}),
  filter(([name, field]) => field.validate), // eslint-disable-line no-unused-vars
  toPairs,
  map(field => ({ ...field, validate: makeValidationRule(field) })),
);

// const defaultValueForType = type => {
//   switch(type = 'string') {
//   case 'number':
//       return 0;
//   case 'string':
//       return '';
//   }
// };

// const getDomainValue = (domainValues, valueId) =>
//   domainValues &&
//   valueId &&
//   reduce((acc, { id, value }) => ({ ...acc, [id]: value }), {}, domainValues)[
//     valueId
//   ];

export const getOneField = fields => name => ({ name, ...fields[name] });
export const getAllFields = compose(
  reduce((acc, [name, field]) => [...acc, { ...field, name }], []),
  toPairs,
);
export const getDefaultValues = compose(
  reduce((acc, [name, field]) => ({ ...acc, [name]: field.defaultValue }), {}),
  toPairs,
);
