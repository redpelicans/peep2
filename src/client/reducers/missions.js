import { compose, fromPairs, map, omit } from 'ramda';
import {
  MISSIONS_LOADED,
  FILTER_MISSIONS_LIST,
  ADD_MISSION,
  MISSION_ADDED,
  MISSION_UPDATED,
  MISSION_DELETED,
  SORT_MISSIONS_LIST,
} from '../actions/missions';

const make = mission => {
  const updatedMission = {
    ...mission,
    typeName: 'mission',
    createdAt: mission.createdAt ? mission.createdAt : undefined,
  };
  if (mission.updatedAt) updatedMission.updatedAt = mission.updatedAt;
  return updatedMission;
};

const makeAll = compose(fromPairs, map(c => [c._id, make(c)]));

const initialState = {
  data: {},
  sort: { by: 'name', order: 'asc' },
  filter: '',
};

const missions = (state = initialState, action) => {
  switch (action.type) {
    case SORT_MISSIONS_LIST: {
      const { by, order } = state.sort;
      const newOrder = by === action.sortBy && order === 'asc' ? 'desc' : 'asc';
      return { ...state, sort: { by: action.sortBy, order: newOrder } };
    }
    case FILTER_MISSIONS_LIST:
      return { ...state, filter: action.filter };
    case MISSIONS_LOADED:
      return { ...state, data: makeAll(action.payload) };
    case ADD_MISSION:
      return { ...state, pending_action: true };
    case MISSION_ADDED:
      return {
        ...state,
        pending_action: false,
        data: {
          ...state.data,
          [action.payload._id]: make(action.payload),
        },
      };
    case MISSION_UPDATED:
      return {
        ...state,
        data: {
          ...state.data,
          [action.payload._id]: make(action.payload),
        },
      };
    case MISSION_DELETED:
      return { ...state, data: omit([action.payload._id])(state.data) };
    default:
      return state;
  }
};

export default missions;
