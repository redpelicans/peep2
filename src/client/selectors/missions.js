export const getFilter = state => state.missions.filter;
export const getSort = state => state.missions.sort;
export const getMissions = state => state.missions.data;
export const getMission = (state, id) => state.missions.data[id];
