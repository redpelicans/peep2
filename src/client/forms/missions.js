import Yup from 'yup';
import { getDefaultValues, getOneValidationSchema, getOneField } from './utils';
import { InputField } from '../fields';
import {
  SelectField,
  AssigneesSelectField,
  CompaniesSelectField,
  WorkerSelectField,
} from '../fields/SelectField';
import { DateField } from '../fields/DateField';
import { MarkDownField } from '../fields/MarkDownField';

const fields = {
  name: {
    label: 'Name',
    required: true,
    component: InputField,
  },
  clientId: {
    label: 'Client',
    component: CompaniesSelectField,
    required: true,
  },
  partner: {
    label: 'Partner',
    component: CompaniesSelectField,
  },
  billedTarget: {
    label: 'Billed Target',
    domainValues: [
      { id: 'partner', value: 'Partner' },
      { id: 'client', value: 'Client' },
    ],
    component: SelectField,
  },
  manager: {
    label: 'Manager',
    component: WorkerSelectField,
    required: true,
  },
  workers: {
    label: 'Workers',
    component: AssigneesSelectField,
  },
  note: {
    label: 'Note',
    component: MarkDownField,
  },
  allowWeekends: {
    label: 'Allow Weekends',
    domainValues: [
      { id: 'allow', value: 'Allow' },
      { id: 'doNotAllow', value: 'Do not Allow' },
    ],
    component: SelectField,
  },
  timesheetUnit: {
    label: 'Timesheet Unit',
    domainValues: [{ id: 'day', value: 'Day' }, { id: 'hour', value: 'Hour' }],
    component: SelectField,
  },
  startDate: {
    label: 'Start Date',
    component: DateField,
  },
  endDate: {
    label: 'End Date',
    component: DateField,
  },
};

export const defaultValues = getDefaultValues(fields);
export const getField = getOneField(fields);
export const getValidationSchema = extend =>
  Yup.object().shape(getOneValidationSchema(fields, extend));
