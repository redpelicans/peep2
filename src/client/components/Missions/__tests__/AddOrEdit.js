import React from 'react';
import { shallow } from 'enzyme';
import Component from '../AddOrEdit';
import { defaultValues } from '../../../forms/missions';
import { mockDate } from '../../../utils/date';

mockDate();

describe('app | components | Mission | component', () => {
  describe('render', () => {
    const props = {
      handleSubmit: jest.fn(),
      values: defaultValues,
      setFieldValue: jest.fn(),
      setFieldTouched: jest.fn(),
      errors: {},
      touched: {},
    };

    it('should be defined', () => {
      // given
      const wrapper = shallow(<Component {...props} />);

      // then
      expect(wrapper).toBeDefined();
    });

    it('should match snapshot', () => {
      // given
      const wrapper = shallow(<Component {...props} />);

      // then
      expect(wrapper).toMatchSnapshot();
    });
  });
});
