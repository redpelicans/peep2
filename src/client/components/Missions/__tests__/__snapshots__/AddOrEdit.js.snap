// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app | components | Mission | component render should match snapshot 1`] = `
<styled.form
  id="missionForm"
  onSubmit={[Function]}
>
  <Styled(FormField)
    errors={Object {}}
    field={
      Object {
        "component": [Function],
        "label": "Name",
        "name": "name",
        "required": true,
        "validate": StringSchema {
          "_blacklist": Set {},
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": Set {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      }
    }
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    touched={Object {}}
    values={
      Object {
        "allowWeekends": "doNotAllow",
        "billedTarget": "client",
        "clientId": undefined,
        "managerId": undefined,
        "name": undefined,
        "note": undefined,
        "partnerId": undefined,
        "timesheetUnit": "day",
      }
    }
  />
  <Styled(FormField)
    errors={Object {}}
    field={
      Object {
        "component": [Function],
        "label": "Client",
        "name": "clientId",
        "required": true,
        "validate": StringSchema {
          "_blacklist": Set {},
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": Set {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      }
    }
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    touched={Object {}}
    values={
      Object {
        "allowWeekends": "doNotAllow",
        "billedTarget": "client",
        "clientId": undefined,
        "managerId": undefined,
        "name": undefined,
        "note": undefined,
        "partnerId": undefined,
        "timesheetUnit": "day",
      }
    }
  />
  <Styled(FormField)
    errors={Object {}}
    field={
      Object {
        "component": [Function],
        "label": "Partner",
        "name": "partnerId",
        "validate": StringSchema {
          "_blacklist": Set {},
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": Set {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      }
    }
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    touched={Object {}}
    values={
      Object {
        "allowWeekends": "doNotAllow",
        "billedTarget": "client",
        "clientId": undefined,
        "managerId": undefined,
        "name": undefined,
        "note": undefined,
        "partnerId": undefined,
        "timesheetUnit": "day",
      }
    }
  />
  <Styled(FormField)
    errors={Object {}}
    field={
      Object {
        "component": [Function],
        "defaultValue": "client",
        "domainValues": Array [
          Object {
            "id": "partner",
            "value": "Partner",
          },
          Object {
            "id": "client",
            "value": "Client",
          },
        ],
        "label": "Billed Target",
        "name": "billedTarget",
        "required": true,
        "validate": StringSchema {
          "_blacklist": Set {},
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": Set {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      }
    }
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    touched={Object {}}
    values={
      Object {
        "allowWeekends": "doNotAllow",
        "billedTarget": "client",
        "clientId": undefined,
        "managerId": undefined,
        "name": undefined,
        "note": undefined,
        "partnerId": undefined,
        "timesheetUnit": "day",
      }
    }
  />
  <Styled(FormField)
    errors={Object {}}
    field={
      Object {
        "component": [Function],
        "label": "Manager",
        "name": "managerId",
        "required": true,
        "validate": StringSchema {
          "_blacklist": Set {},
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": Set {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      }
    }
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    touched={Object {}}
    values={
      Object {
        "allowWeekends": "doNotAllow",
        "billedTarget": "client",
        "clientId": undefined,
        "managerId": undefined,
        "name": undefined,
        "note": undefined,
        "partnerId": undefined,
        "timesheetUnit": "day",
      }
    }
  />
  <Styled(FormField)
    errors={Object {}}
    field={
      Object {
        "component": [Function],
        "label": "Note",
        "name": "note",
        "validate": StringSchema {
          "_blacklist": Set {},
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": Set {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      }
    }
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    touched={Object {}}
    values={
      Object {
        "allowWeekends": "doNotAllow",
        "billedTarget": "client",
        "clientId": undefined,
        "managerId": undefined,
        "name": undefined,
        "note": undefined,
        "partnerId": undefined,
        "timesheetUnit": "day",
      }
    }
  />
  <Styled(FormField)
    errors={Object {}}
    field={
      Object {
        "component": [Function],
        "defaultValue": "doNotAllow",
        "domainValues": Array [
          Object {
            "id": "allow",
            "value": "Allow",
          },
          Object {
            "id": "doNotAllow",
            "value": "Do not Allow",
          },
        ],
        "label": "Allow Weekends",
        "name": "allowWeekends",
        "validate": StringSchema {
          "_blacklist": Set {},
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": Set {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      }
    }
    setFieldTouched={[Function]}
    setFieldValue={[Function]}
    touched={Object {}}
    values={
      Object {
        "allowWeekends": "doNotAllow",
        "billedTarget": "client",
        "clientId": undefined,
        "managerId": undefined,
        "name": undefined,
        "note": undefined,
        "partnerId": undefined,
        "timesheetUnit": "day",
      }
    }
  />
</styled.form>
`;
